### DISCUSSION TITLE
ファイル不足の表示機能の提案

### DISCUSSION THREAD

**Discussion by Nekokan** [exch-bms2/beatoraja#667] [Ideas]
フォーラムが新設されたことを聞いて試し利用も兼ねて。

非常に古い話でBM98やDelight Delight Reduplicationではダイアログボックス、nazoBMPlayでは演奏画面中に表示されるのですが、
BMSに必要なリソース、すなわち#WAVや#BMPが存在しない場合にその事実とファイル名を知らせる機能があって大変便利でした。
この機能の実装について是非やアイデアなどを検討してほしいと思いました。
（もし既にあったとしたらすごく恥ずかしい）

ちなみになぜこの提案をするに至ったかといいますと、拙作BOFXVIIパッケージにてファイル不足の指摘が上がっているためです。 [doubledepth 2021-12](https://hitkey.nekokan.dyndns.info/diary2112.php#D211212B)
以下のような声もあり、気づかないうちに不完全な形で作品を遊んでいるケースが少なくないと思った次第です。
[https://twitter.com/AYYYhaz/status/1464376509020491776?s=20](https://twitter.com/AYYYhaz/status/1464376509020491776)

もしかしたらどなたが確認ツールを開発しているかもしれませんが恐らくその対象は曲作者側で、ユーザーやパッケージ作者が作品の数だけチェックするのは現実的ではなさそうです。

**Comment by exch-bms2** [comment]
リソース不足ログを収集するのは難しくないと思います。
あとはユーザーへの情報開示方法についてですが、何かアイデアありますか？

思いついたものだと、
・テキストログにそのまま出力(目的のログを探すのが大変？)
・songdata.dbにlogカラムを用意して、そこに出力する(songdata.dbのサイズ/パフォーマンスへの影響は？)
・対称のbmsファイルと同じ階層にログファイルを出力(間違ってログファイルも含められて作品登録されそう)

**Reply by SackMagiclight** [reply]
songdata.dbにlogカラムを用意して、そこに出力する（初回DB登録時やフォルダリロード時にログDB格納）に一票いれたいですねー  
スキン定義用意すれば選曲・プレー画面に出せると思うので。

songdata.dbにはロード時欠損していたかの情報（Boolean）、ログに欠損ファイル詳細のようなハイブリッドもありかもしれません。

**Reply by Nekokan** [reply]
スキンで表示をコントロールできるのですか。それは面白そうですね。
それは一方でスキン次第で表示されないということにもなりそうですが、恐らくそれは欲張りなのだろうと理解しておきます。

個人的に重視したいのは　(1)演奏開始前までに (2)プレイヤーが容易に把握できるように　なっているとうれしいです。ファイル不足なら遊ばない、再DLするという選択ができるためです。たとえ遊んだ場合でも、どのタイミングでファイル不足に気づくかで作品の評価にかかわってくる可能性もあります。

ただ昔のソフトのようにWindowsのダイアログボックスでゲームを一時中断させてしまうのも考えものです。あるいはbm98はファイルが足りないと再生自体しない仕様でしたが、さすがにそこまではしない方がよいと思いました。ファイル不足を承知の上で遊ぶのはユーザーの自由にしてよいと思います。

**Reply by exch-bms2** [reply]
(1)演奏開始前までに

楽曲更新時にリソースの存在も含めたパースをかけるのは結構重くならないかな？という懸念があります。
実装してみないと分かりませんが。

(2)プレイヤーが容易に把握できるように

選曲時にリソース存在率をスキン側で取得できるようにするとか、閾値以下の楽曲にマークをつけるとか…

**Reply by kasakon555** [reply]
リソース不足を確認できるのはいいですね！
スキン定義ができればこれまでBGAファイルが定義されていて存在しなかった場合に真っ黒になっていたのが汎用BGAなどを表示できるようになるのでかなり便利そうです。
BGAの拡張子などでスキン側に表示するのも面白そうかなと個人的に思いました。
一定割合（選択式？）のリソース不足で切り替わる定義があれば差分を入れ間違えたときにも役立ちそうです。

**Comment by exch-bms2** [comment]
songdata.db更新時にリソースチェックをかけるとして、

・songdata.dbのカラム名とその型式 (とりあえずlog:Stringでjsonで結果を突っ込む？)
・スキン側への提供方法

のアイデア出しですかね。
